ext/standard/array.c

```diff
@@ -6389,3 +6389,155 @@ PHP_FUNCTION(array_combine)
 	} ZEND_HASH_FOREACH_END();
 }
 /* }}} */
+
+static inline void php_array_until(zval *return_value, HashTable *htbl, zend_fcall_info fci, zend_fcall_info_cache fci_cache, int stop_value) /* {{{ */
+{
+	zval args[1];
+	zend_bool have_callback = 0;
+	zval *operand;
+	zval retval;
+	int result;
+
+	if (zend_hash_num_elements(htbl) == 0) {
+		RETURN_BOOL(!stop_value);
+	}
+
+	if (ZEND_FCI_INITIALIZED(fci)) {
+		have_callback = 1;
+		fci.retval = &retval;
+		fci.param_count = 1;
+		fci.params = args;
+	}
+
+	ZEND_HASH_FOREACH_VAL(htbl, operand) {
+		if (have_callback) {
+			zend_bool retval_true;
+			ZVAL_COPY(&args[0], operand);
+
+			/* Treat the operand like an array of size 1  */
+			result = zend_call_function(&fci, &fci_cache);
+			zval_ptr_dtor(operand);
+			if (result == FAILURE) {
+				return;
+			}
+			retval_true = zend_is_true(&retval);
+			zval_ptr_dtor(&retval);
+			/* The user-provided callback rarely returns refcounted values. */
+			if (retval_true == stop_value) {
+				RETURN_BOOL(stop_value);
+			}
+		} else {
+			if (zend_is_true(operand) == stop_value) {
+				RETURN_BOOL(stop_value);
+			}
+		}
+	} ZEND_HASH_FOREACH_END();
+
+	RETURN_BOOL(!stop_value);
+}
+/* }}} */
+
+typedef struct {
+	zval                   *obj;
+	int                    stop_value;
+	int                    result;
+	int                    found;
+	zend_long              has_callback;
+	zend_fcall_info        fci;
+	zend_fcall_info_cache  fcc;
+} php_iterator_until_info;
+
+static int php_traversable_func_until(zend_object_iterator *iter, void *puser) /* {{{ */
+{
+	zend_fcall_info fci;
+	zval retval;
+	php_iterator_until_info *until_info = (php_iterator_until_info*) puser;
+	int result;
+	zend_bool stop;
+
+	fci = until_info->fci;
+	if (ZEND_FCI_INITIALIZED(fci)) {
+		zval *operand = iter->funcs->get_current_data(iter);
+		fci.retval = &retval;
+		fci.param_count = 1;
+		/* Use the operand like an array of size 1 */
+		fci.params = operand;
+		fci.retval = &retval;
+		Z_TRY_ADDREF_P(operand);
+		result = zend_call_function(&fci, &until_info->fcc);
+		zval_ptr_dtor(operand);
+		if (result == FAILURE) {
+			until_info->result = FAILURE;
+			return ZEND_HASH_APPLY_STOP;
+		}
+
+		stop = zend_is_true(&retval) == until_info->stop_value;
+		zval_ptr_dtor(&retval);
+	} else {
+		stop = zend_is_true(iter->funcs->get_current_data(iter)) == until_info->stop_value;
+	}
+	if (stop) {
+		until_info->found = 1;
+		return ZEND_HASH_APPLY_STOP;
+	}
+	return ZEND_HASH_APPLY_KEEP;
+}
+/* }}} */
+
+static inline void php_iterable_until(INTERNAL_FUNCTION_PARAMETERS, int stop_value) /* {{{ */
+{
+	zval *input;
+	zend_fcall_info fci = empty_fcall_info;
+	zend_fcall_info_cache fci_cache = empty_fcall_info_cache;
+
+	ZEND_PARSE_PARAMETERS_START(1, 2)
+		Z_PARAM_ZVAL(input)
```

Comment:
Do we not have ` Z_PARAM_*` for `iterable`?

Reply:
I checked and didn't find anything - https://php.net/iterator_map takes an iterator but not an iterable. I didn't see any .stub.php files using iterable (see build/gen_stubs.php change)

Reply:
Note that I opened PR #6064: "Add Z_PARAM_ITERABLE and co" to address this.

Reply:
This should now be able to use `Z_PARAM_ITERABLE` as the PR merged.